### DISCUSSION TITLE
Trying to convert the APRS_Position_LoRa.ino example to STM32WL for Ebyte's E77 STM32WLE5CCU6 board

### DISCUSSION THREAD

**Discussion by radiohound** [jgromes/RadioLib#1323] [Q&A] [answered]
Hello,

I am trying to get the APRS_Position_LoRa.ino going on a STM32WLE board. Hardware is [_(red: fixed link formatting)_](https://www.cdebyte.com/products/E77-400MBL-01#Specification) As far as the serial responses, it acts like its transmitting. But I never get a reception from my iGates.

I am currently compiling for:
```
[env:lora_e5_dev_board]
platform = ststm32
board = lora_e5_dev_board
board_build.mcu = stm32wle5cc ;stm32wle5ccu6 on Ebyte E77 433mhz board
```

But I have also tried compiling for a nucleo_wl55jc, with the same results. Can anyone tell what I am doing wrong?
Thank you, Walter

------------------------------------------------------------------------------------
```cpp
// include the library
#include <Arduino.h>
#include <RadioLib.h>

// no need to configure pins, signals are routed to the radio internally
STM32WLx radio = new STM32WLx_Module();

APRSClient aprs(&radio);

// set RF switch configuration for Nucleo WL55JC1
// NOTE: other boards may be different!
//       Some boards may not have either LP or HP.
//       For those, do not set the LP/HP entry in the table.
static const uint32_t rfswitch_pins[] =
                         {PC3,  PC4,  PC5, RADIOLIB_NC, RADIOLIB_NC};
static const Module::RfSwitchMode_t rfswitch_table[] = {
  {STM32WLx::MODE_IDLE,  {LOW,  LOW,  LOW}},
  {STM32WLx::MODE_RX,    {HIGH, HIGH, LOW}},
  {STM32WLx::MODE_TX_LP, {HIGH, HIGH, HIGH}},
  {STM32WLx::MODE_TX_HP, {HIGH, LOW,  HIGH}},
  END_OF_MODE_TABLE,
};

void setup() {
  Serial.begin(9600);
  pinMode(PB4, OUTPUT);     // Set LED1 as output
  pinMode(PB3, OUTPUT);     // Set LED2 as output
  digitalWrite(PB4, 0);     // Turn off LED1
  digitalWrite(PB3, 0);     // Turn off LED2
 
  radio.setRfSwitchTable(rfswitch_pins, rfswitch_table);
  
  // set RF switch control configuration
  // this has to be done prior to calling begin()
  //radio.setRfSwitchTable(rfswitch_pins, rfswitch_table);

  // initialize STM32WL with default settings, except frequency
  Serial.print(F("[STM32WL] Initializing ... "));
  //int state = radio.begin(433.775);
  // frequency:                   433.775 MHz
  // bandwidth:                   125 kHz
  // spreading factor:            12
  // coding rate:                 4/5
  int state = radio.begin(433.775, 125, 12, 5);

  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true) { delay(10); }
  }

  // initialize APRS client
  Serial.print(F("[APRS] Initializing ... "));
  // symbol:                      '>' (car)
  // callsign                     "K6ATV"
  // SSID                         1
  char source[] = "K6ATV";
  state = aprs.begin('>', source, 1);

  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true) { delay(10); }
  }

  // set appropriate TCXO voltage for Nucleo WL55JC1
  state = radio.setTCXO(1.7);
  Serial.print(F("Set TCXO voltage ... "));
  
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true) { delay(10); }
  }
}

  // counter to keep track of transmitted packets
  int count = 0;

void loop() {
  //**************************************************

  Serial.print(F("[APRS] Sending position ... "));
  
  // send a location with message and timestamp
  // SSID is set to 1, as APRS over LoRa uses WIDE1-1 path by default
  char destination[] = "GPS";
  char latitude[] = "3011.67N";
  char longitude[] = "12135.96W";
  char message[] = "I'm here!";
  char timestamp[] = "093045z";
  digitalWrite(PB3, 1);
  //int state = aprs.sendPosition(destination, 1, latitude, longitude, message, timestamp);
  int state = aprs.sendPosition(destination, 1, latitude, longitude, message);
  digitalWrite(PB3, 0);
  delay(5000);

  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
  }

  Serial.print(F("[Mic-E] Sending position ... "));
  // you can also send Mic-E encoded messages
  digitalWrite(PB3, 1);
  state |= state = aprs.sendMicE(30.1945, -121.6000, 120, 10, RADIOLIB_APRS_MIC_E_TYPE_EN_ROUTE);
  digitalWrite(PB3, 0);
  delay(5000);

  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
  }

  // wait for 30 seconds before transmitting again
  delay(30000);
}
```

**Comment by StevenCellist** [comment]
Hi @radiohound, a few comments:
- Please make sure you use Markdown formatting when pasting your code, such that we can read it. The same goes for the link to the module you linked: it wasn't rendered. I just fixed it for you, but please do this yourself the next time you insert a code block or link (the buttons are in the same top bar as the Write/Preview boxes when you type a message).
- Am I right that you are using the STM32duino core? It isn't specifically clear from your `platformio.ini`, but as far as I can see, STM32duino would be the default. Can you confirm this please?
- Have you tried the examples that are specific for the STM32WLx series? Just a plain transmit and/or receive.
- You may want to enable the BASIC debug level (refer to the wiki) to get some insight in whether there is something happening under the hood that is unexpected. Not many people use the STM32WL boards, and even fewer would use it for APRS.

**Reply by radiohound** [reply]
Thank you for your reply @StevenCellist . 

Thank you for the Markdown help. 
Yes, I am using the STM32duino core.
I have now tried the send and receive examples, and can not get data over the modules. I get timeouts waiting for the reception on the RX side. 

On the RX side I have run basic debug, and get the following result:

```
RadioLib Info
Version:  "7.0.0.0"
Platform: "Arduino STM32 (official)"
Compiled: "Nov 18 2024" "06:23:59"
RLB_DBG: Found SX126x: RADIOLIB_SX126X_REG_VERSION_STRING:
RLB_DBG: 00000320: 53 58 31 32 36 31 20 54 4b 46 20 31 41 31 31 00  SX1261 TKF 1A11.
RLB_DBG: 
RLB_DBG: M      SX126x
success!
success!
[STM32WL] Waiting for incoming transmission ... RLB_DBG: Timeout in 409 ms
```

On the TX side I have run basic debug, and get the following result:

```
RadioLib Info
Version:  "7.0.0.0"
Platform: "Arduino STM32 (official)"
Compiled: "Nov 18 2024" "06:23:59"
RLB_DBG: Found SX126x: RADIOLIB_SX126X_REG_VERSION_STRING:
RLB_DBG: 00000320: 53 58 31 32 36 31 20 54 4b 46 20 31 41 31 30 00  SX1261 TKF 1A10.
RLB_DBG: 
RLB_DBG: M      SX126x
success!
success!
[STM32WL] Transmitting packet ... RLB_DBG: Timeout in 988 ms
success!
[STM32WL] Datarate:     615.38 bps
[STM32WL] Transmitting packet ... RLB_DBG: Timeout in 988 ms
success!
[STM32WL] Datarate:     618.56 bps
[STM32WL] Transmitting packet ... RLB_DBG: Timeout in 988 ms
success!
[STM32WL] Datarate:     618.56 bps
```

**Reply by jgromes** [reply]
@radiohound First of all, it looks like you are using an older version of the library - 7.0.0 while the latest release is 7.1.0, so the first thing to try is to update.

Second, the debug output you have posted does not seem to match the code shown above - are you running the default examples now? I would also suggest using interrupt-driven reception as opposed to blocking.

Finally, STM32WL has a bit more complex RF switching than most other modules, so I would suggest to double-check that.

**Reply by radiohound** [reply]
When I try to compile STM32WLx_Transmit_Interrupt with the latest library 7.1.0, I get an error:
(I also get the same error with version 7.0.0)

```
src/main.cpp:71:23: error: 'setFlag' was not declared in this scope
   71 |   radio.setDio1Action(setFlag);
```

**Comment by radiohound** [comment]
I am convinced you are right, that my problem is RF switch pins, or possibly some setting like TCXO or HSE. But I am confused about how the RF Switch pins function. These switch pins don't seem to be chip specific. They seem to be board specific. For instance, in stmduino:

For generic node:
```
from https://github.com/stm32duino/Arduino_Core_STM32/blob/main/variants/STM32WLxx/WL54CCU_WL55CCU_WLE4C(8-B-C)U_WLE5C(8-B-C)U/variant_GENERIC_NODE_SE_TTI.h
#define LORAWAN_RFSWITCH_PINS           PA0,PA1,PB8
#define LORAWAN_RFSWITCH_PIN_COUNT      3
#define LORAWAN_RFSWITCH_OFF_VALUES     LOW,LOW,LOW
#define LORAWAN_RFSWITCH_RX_VALUES      HIGH,LOW,HIGH
#define LORAWAN_RFSWITCH_RFO_LP_VALUES  HIGH,HIGH,HIGH
#define LORAWAN_RFSWITCH_RFO_HP_VALUES  LOW,HIGH,HIGH
```

For RAK3172 Module:
```
https://github.com/stm32duino/Arduino_Core_STM32/blob/main/variants/STM32WLxx/WL54CCU_WL55CCU_WLE4C(8-B-C)U_WLE5C(8-B-C)U/variant_RAK3172_MODULE.h
#define LORAWAN_RFSWITCH_PINS           PB8, PC13
#define LORAWAN_RFSWITCH_PIN_COUNT      2
#define LORAWAN_RFSWITCH_OFF_VALUES     LOW, LOW
#define LORAWAN_RFSWITCH_RX_VALUES      HIGH, LOW
#define LORAWAN_RFSWITCH_RFO_LP_VALUES  LOW, HIGH
#define LORAWAN_RFSWITCH_RFO_HP_VALUES  LOW, HIGH
```

I am trying to use a STM32WLE5CC module in a stand alone application, with no external control. Are these pins controlling the SX1261 (inside the STM32WLE5 chip)? Or are these pins meant for outside control of the module? If the second part is true, I can understand why it is board specific. But if these pins are connected internally to the SX1261, then I do not understand why these pins appear to be board specific.

Thanks for any ideas you may have.

**Reply by jgromes** [reply]
The RF switch is external to the SX126x, which has separate RF out and RF in pins. In tthe case of the STM32WL, the SX126x is embedded within the package (probably on the same silicon die considering its SPI is hooked up directly to the STM32's AHB bus). If you take a look at the STM32WL55xx datasheet, specifically the pinout section, you will find that it also has separate RF in and RF out pins. What's more, it even has a separate high power and low power outputs. So the RF switching pins are board-specific.

**Reply by radiohound** [reply]
Thank you @jgromes , that made the lightbulb in my head turn on, and it now makes sense to me. I appreciate the extra help!

I wasn't seeing much hardware outside the chip, but I guess the 6 pin device is a switch. I think it's PA6 and PA7, because those are not brought out to the boards IO pins. But I have a question out to the manufacturer. Once I figure it out or find out, I'll update this for the E77 board.

![image](https://github.com/user-attachments/assets/92062089-31c7-4057-96d7-42c180c58426)

**Comment by K4KDR** [comment]
I am not skilled enough to know if my situation is applicable to the discussion here, but wanted to share my config that works perfectly for the **STM32WL55JC1** &  **JC2**:

```
//-----------------------------------------------------------------------------
// set up radio module
// no need to configure pins, signals are routed to the radio internally
STM32WLx radio = new STM32WLx_Module();

// set RF switch configuration for Nucleo WL55JC2
// NOTE: other boards may be different!
//       Some boards may not have either LP or HP.
//       For those, do not set the LP/HP entry in the table.
static const uint32_t rfswitch_pins[] =
                         {PC3,  PC4,  PC5, RADIOLIB_NC, RADIOLIB_NC};
static const Module::RfSwitchMode_t rfswitch_table[] = {
  {STM32WLx::MODE_IDLE,  {LOW,  LOW,  LOW}},
  {STM32WLx::MODE_RX,    {HIGH, HIGH, LOW}},
  {STM32WLx::MODE_TX_LP, {HIGH, HIGH, HIGH}},
  {STM32WLx::MODE_TX_HP, {HIGH, LOW,  HIGH}},
  END_OF_MODE_TABLE,
};

//-----------------------------------------------------------------------------
```

**Comment by radiohound** [comment]
Thank you K4KDR,

I have a STM32WL55JC2 coming in the mail, so I can try it with some of the
existing code, using your RF configs. I gather there are different settings
on the WLE series. Apparently multiple different settings.... I may have to
ask EByte, the manufacturer of my module for the specific info for my E77
Ebyte module. You would think they would publish the info, like Seeed
Studio does for theirs, but I can't seem to find it anywhere (yet).

For instance, the WIO-E5 (a STM32WLE5JC) uses only 2 switches:
Wio-E5 module ONLY transmits through RFO_HP:
   Receive: PA4=1, PA5=0
   Transmit(high output power, SMPS mode): PA4=0, PA5=1

Walter

**Comment by radiohound** [comment, marked as answer]
Ok, I was able to get the modules to transmit and receive using the example code. For this module, there are only 2 outputs for the switch, but I use the third one to turn on an LED for transmitting.  Thanks for everyones help.

```
// set RF switch configuration for EBytes E77 board
// PB3 is an LED - activates while transmitting
// NOTE: other boards may be different!
//       Some boards may not have either LP or HP.
//       For those, do not set the LP/HP entry in the table.
static const uint32_t rfswitch_pins[] =
                         {PA6,  PA7,  PB3, RADIOLIB_NC, RADIOLIB_NC};
static const Module::RfSwitchMode_t rfswitch_table[] = {
  {STM32WLx::MODE_IDLE,  {LOW,  LOW,  LOW}},
  {STM32WLx::MODE_RX,    {LOW, HIGH, LOW}},
  {STM32WLx::MODE_TX_LP, {HIGH, LOW, HIGH}},
  {STM32WLx::MODE_TX_HP, {HIGH, LOW,  HIGH}},
  END_OF_MODE_TABLE,
};
```
